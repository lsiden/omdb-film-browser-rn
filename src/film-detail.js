import React from "react"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import { Text, View, FlatList, Linking, Image } from "react-native"
import cuid from "cuid"

import { viewList } from "./actions"
import { CloseButton } from "components/close-button"
import { detailStyles } from "./styles"

function renderItem(item) {
  const { text, cond = true, url } = item

  if (!cond) {
    return null
  }

  const props = {
    text,
    style: detailStyles.itemStyle,
  }

  if (url) {
    props.onPress = () => {
      Linking.openURL(url)
    }
  }
  return <Text {...props}>{text}</Text>
}

const keyExtractor = item => cuid.slug()

export class filmDetail extends React.Component {
  constructor(props) {
    const { imdbID } = props.filmDetails
    super(props)
    this.imdbUrl = `https://www.imdb.com/title/${imdbID}`
    this.dispatchViewList = props.dispatchViewList.bind(this)
  }

  renderTitle() {
    const { dispatchViewList, filmSummary } = this.props
    return (
      <View style={detailStyles.headerStyle}>
        <Text style={detailStyles.titleStyle}>{filmSummary.Title}</Text>
        <CloseButton onPress={dispatchViewList} />
      </View>
    )
  }

  renderDetails() {
    const { filmSummary, filmDetails } = this.props
    const items = [
      { text: filmSummary.Year },
      { text: `Directed by $${filmDetails.Director}` },
      { text: `Written by ${filmDetails.Writer}` },
      { text: `Cast: ${filmDetails.Actors}` },
      { text: `Language: ${filmDetails.Language}` },
      { text: `Awards: {filmDetails.Awards}`, cond: filmDetails.Awards },
      { text: `Run Time: ${filmDetails.Runtime}` },
      { text: `IMDB Rating: ${filmDetails.imdbRating}/10` },
      { text: `Box Office: ${filmDetails.BoxOffice}` },
      {
        text: "Official Website",
        url: filmDetails.Website,
        cond: filmDetails.Website,
      },
      {
        text: "IMDB page",
        url: this.imdbUrl,
      },
    ]
    return (
      <FlatList
        data={items}
        renderItem={renderItem}
        keyExtractor={keyExtractor}
      />
    )
  }

  render() {
    const { filmSummary } = this.props
    return (
      <View style={detailStyles.wrapperStyle}>
        {this.renderTitle()}
        <Image Source={filmSummary.Poster} />
        {this.renderDetails()}
      </View>
    )
  }
}

filmDetail.propTypes = {
  filmSummary: PropTypes.object.isRequired,
  filmDetails: PropTypes.object,
  dispatchViewList: PropTypes.func.isRequired,
}

filmDetail.defaultProps = {
  filmDetails: {},
}

export default connect(
  state => ({
    filmSummary: state.filmSummary,
    filmDetails: state.filmDetails,
  }),
  dispatch => ({
    dispatchViewList: () => dispatch(viewList()),
  })
)(filmDetail)
