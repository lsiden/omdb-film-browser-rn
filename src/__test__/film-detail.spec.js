import React from "react"
import renderer from "react-native-test-utils"

import { filmDetail as FilmDetail } from "film-detail"

const films = require("./__fixture__/films.json").Search
const filmDetails = require("./__fixture__/film-detail.json")
const filmSummary = films[0]

let dispatchViewList

const defaultProps = () => {
  dispatchViewList = jest.fn()
  return {
    filmSummary,
    filmDetails,
    dispatchViewList,
  }
}

const createWrapper = (props = {}) =>
  renderer(<FilmDetail {...{ ...defaultProps(), ...props }} />)

it("renders a title", () => {
  const wrapper = createWrapper()
  expect(wrapper.toJSON()).toMatchSnapshot()
})
