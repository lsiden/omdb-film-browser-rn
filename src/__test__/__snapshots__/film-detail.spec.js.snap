// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a title 1`] = `
<View>
  <View
    style={
      Object {
        "alignContent": "flex-start",
        "alignItems": "baseline",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 24,
          }
        }
      >
        ‚óÅ
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "flex": 2,
          "fontSize": 20,
          "fontWeight": "bold",
          "marginBottom": 5,
          "textAlign": "center",
        }
      }
    >
      The Battle of Algiers
    </Text>
  </View>
  <Image
    resizeMode="contain"
    source={
      Object {
        "uri": "https://images-na.ssl-images-amazon.com/images/M/MV5BYzliNzI5NzgtNDZmNy00NGU4LTg2OTQtMDEzZTRiNDU3ZTI1XkEyXkFqcGdeQXVyMDUyOTUyNQ@@._V1_SX300.jpg",
      }
    }
    style={
      Object {
        "height": 300,
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  />
  <RCTScrollView
    data={
      Array [
        Object {
          "text": "1966",
        },
        Object {
          "text": "Directed by Gillo Pontecorvo",
        },
        Object {
          "text": "Written by Franco Solinas, Franco Solinas (story), Gillo Pontecorvo (story)",
        },
        Object {
          "text": "Cast: Jean Martin, Yacef Saadi, Brahim Hadjadj, Tommaso Neri",
        },
        Object {
          "text": "Language: French, Arabic, English",
        },
        Object {
          "cond": true,
          "text": "Awards: Nominated for 3 Oscars. Another 9 wins & 5 nominations.",
        },
        Object {
          "text": "Run Time: 121 min",
        },
        Object {
          "text": "IMDB Rating: 8.1/10",
        },
        Object {
          "text": "Box Office: $55,908",
        },
        Object {
          "cond": "http://rialtopictures.com/battle.html",
          "text": "Official Website",
          "url": "http://rialtopictures.com/battle.html",
        },
        Object {
          "text": "IMDB page",
          "url": "https://www.imdb.com/title/tt0058946",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          1966
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          Directed by Gillo Pontecorvo
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          Written by Franco Solinas, Franco Solinas (story), Gillo Pontecorvo (story)
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          Cast: Jean Martin, Yacef Saadi, Brahim Hadjadj, Tommaso Neri
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          Language: French, Arabic, English
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          Awards: Nominated for 3 Oscars. Another 9 wins & 5 nominations.
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          Run Time: 121 min
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          IMDB Rating: 8.1/10
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          Box Office: $55,908
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          onPress={[Function]}
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          Official Website
        </Text>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
